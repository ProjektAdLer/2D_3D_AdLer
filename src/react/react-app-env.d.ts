/// <reference types="react-scripts" />

// Images
declare module '*.jpg';
declare module '*.png';
declare module '*.env';

// 3D types
declare module '*.glb';
declare module '*.stl';

declare module '*.jpeg';
declare module '*.gif';
declare module '*.mp4';
