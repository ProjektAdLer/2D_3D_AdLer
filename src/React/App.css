body {
  margin: 0;
  height: 100vh;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

div{
  width: 100%;
  height: 100%;
}

section{
  display: inline;
  position: fixed;
  top: 0;
  left: 0;
  width: 120px;
  height: 100vh;
  background-color: #fff;
}

button{
  width: 100px;
  margin: 10px;
  z-index: 5;
  position: fixed;
  left: 0;
  top: 0;
}

canvas {
  width: calc(100% - 100px);
  height: 100%;
  margin-left: 100px;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New",
    monospace;
}

.modal {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background: rgba(0, 0, 0, 0.5);
  display: flex;
  align-items: center;
  justify-content: center;
}

.modal-content {
  width: 1000px;
  background-color: #fff;
}

.modal-header,
.modal-footer {
  padding: 10px;
  height: 60px;
}

.modal-title {
  margin: 0;
}

.modal-body {
  padding: 10px;
  border-top: 1px solid #eee;
  border-bottom: 1px solid #eee;
}
