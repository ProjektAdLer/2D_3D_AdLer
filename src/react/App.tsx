import React from 'react';
import './App.css';
import BabylonCanvas from './components/BabylonCanvas';

function App() {
	return (
		<React.Fragment>
			<BabylonCanvas style={{ width: '100%', position: 'absolute' }} />
		</React.Fragment>
	);
}

export default App;
