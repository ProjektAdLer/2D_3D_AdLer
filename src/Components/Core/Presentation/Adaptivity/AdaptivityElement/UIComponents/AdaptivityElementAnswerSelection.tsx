import { useCallback, useEffect, useState } from "react";
import { AdaptivityQuestion } from "../AdaptivityElementViewModel";
import StyledButton, {
  StyledButtonColor,
} from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledButton";
import { useTranslation } from "react-i18next";

function getIsAnyAnswerSelected(question: AdaptivityQuestion): boolean {
  return question.questionAnswers.some((a) => a.isSelected === true);
}

export default function AdaptivityElementAnswerSelection({
  question,
  submitSelection,
  closeSelection,
  setHeaderText,
}: {
  question: AdaptivityQuestion;
  submitSelection: () => void;
  closeSelection: () => void;
  setHeaderText: (headerText: string) => void;
}) {
  const [answerColors, setAnswerColors] = useState<StyledButtonColor[]>([]);
  const [isAnyAnswerSelected, setIsAnyAnswerSelected] = useState<boolean>(
    getIsAnyAnswerSelected(question)
  );

  const { t: translate } = useTranslation("learningElement");
  const [text] = useState<string>(
    (question.isMultipleChoice
      ? translate("multipleChoiceSelected")
      : translate("singleChoiceSelected")) + question.questionText
  );

  useEffect(() => {
    setHeaderText(text);
    setAnswerColors(question.questionAnswers.map(() => "default"));
  }, [setHeaderText, question.questionAnswers, text]);

  const onAnswerClicked = useCallback(
    (index: number) => {
      const newColors = question.questionAnswers.map((answer, i) => {
        if (index === i) {
          // switch selection status of clicked answer
          answer.isSelected = !answer.isSelected;
        } else if (!question.isMultipleChoice) {
          // deselect all other answers if question is single choice
          answer.isSelected = false;
        }

        return answer.isSelected ? "highlight" : "default";
      });

      setIsAnyAnswerSelected(getIsAnyAnswerSelected(question));

      setAnswerColors(newColors);
    },
    [question]
  );

  return (
    <div className="flex flex-col w-full p-2 m-auto lg:justify-between lg:flex-row">
      <div className="flex flex-wrap justify-start gap-4">
        {question.questionAnswers.map((answer, index) => (
          <StyledButton
            key={answer.answerIndex}
            shape="freefloatcenter"
            onClick={() => {
              onAnswerClicked(index);
            }}
            color={answerColors[index]}
          >
            <p className="text-sm">{answer.answerText}</p>
          </StyledButton>
        ))}
      </div>

      <div className="flex items-end justify-end gap-2 pt-8 font-bold lg:w-1/6">
        <StyledButton
          className="box-border "
          shape="freefloatcenter"
          onClick={() => {
            submitSelection();
          }}
          disabled={!isAnyAnswerSelected}
        >
          <p className="text-sm">
            {question.isMultipleChoice
              ? translate("submitAnswers")
              : translate("submitAnswer")}
          </p>
        </StyledButton>
      </div>
    </div>
  );
}
