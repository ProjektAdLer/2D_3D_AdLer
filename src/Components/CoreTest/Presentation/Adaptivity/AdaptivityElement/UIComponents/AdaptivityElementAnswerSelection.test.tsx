import { render } from "@testing-library/react";
import AdaptivityElementAnswerSelection from "../../../../../Core/Presentation/Adaptivity/AdaptivityElement/UIComponents/AdaptivityElementAnswerSelection";
import React from "react";

describe("AdaptivityElementAnswerSelection", () => {
  test("should render", () => {
    const { container } = render(
      <AdaptivityElementAnswerSelection
        question={{
          questionID: 0,
          questionText: "testQuestionText",
          isMultipleChoice: false,
          difficulty: 0,
          isCompleted: false,
          isRequired: false,
          questionAnswers: [
            {
              answerIndex: 0,
              answerText: "testAnswerText",
              isSelected: false,
              isCorrect: false,
            },
          ],
          hints: [],
        }}
        setHeaderText={() => {}}
        submitSelection={() => {}}
        closeSelection={() => {}}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  // ANF-ID: [EWE0008]
  test("submit button is deactive if no answer is selected if question is single choice", () => {
    const { getByText } = render(
      <AdaptivityElementAnswerSelection
        question={{
          questionID: 0,
          questionText: "testQuestionText",
          isMultipleChoice: false,
          difficulty: 0,
          isCompleted: false,
          isRequired: false,
          questionAnswers: [
            {
              answerIndex: 0,
              answerText: "testAnswerText",
              isSelected: false,
              isCorrect: false,
            },
          ],
          hints: [],
        }}
        setHeaderText={() => {}}
        submitSelection={() => {}}
        closeSelection={() => {}}
      />,
    );
    expect(getByText("submitAnswer").closest("button")).toBeDisabled();
  });

  // ANF-ID: [EWE0009]
  test("submit button is deactive if no answer is selected if question is multiple choice", () => {
    const { getByText } = render(
      <AdaptivityElementAnswerSelection
        question={{
          questionID: 0,
          questionText: "testQuestionText",
          isMultipleChoice: true,
          difficulty: 0,
          isCompleted: false,
          isRequired: false,
          questionAnswers: [
            {
              answerIndex: 0,
              answerText: "testAnswerText",
              isSelected: false,
              isCorrect: false,
            },
          ],
          hints: [],
        }}
        setHeaderText={() => {}}
        submitSelection={() => {}}
        closeSelection={() => {}}
      />,
    );
    expect(getByText("submitAnswers").closest("button")).toBeDisabled();
  });

  // ANF-ID: [EWE0008, EWE0009]
  test("click on submit button should call submitSelection", () => {
    const submitSelection = jest.fn();
    const { getByText } = render(
      <AdaptivityElementAnswerSelection
        question={{
          questionID: 0,
          questionText: "testQuestionText",
          isMultipleChoice: false,
          difficulty: 0,
          isCompleted: false,
          isRequired: false,
          questionAnswers: [
            {
              answerIndex: 0,
              answerText: "testAnswerText",
              isSelected: true, // must be true to enable submit button
              isCorrect: false,
            },
          ],
          hints: [],
        }}
        setHeaderText={() => {}}
        submitSelection={submitSelection}
        closeSelection={() => {}}
      />,
    );

    getByText("submitAnswer").click();

    expect(submitSelection).toHaveBeenCalledTimes(1);
  });

  // ANF-ID: [EWE0008, EWE0009]
  test("click on answer button inverts its isSelected flag", () => {
    const answer = {
      answerIndex: 0,
      answerText: "testAnswerText",
      isSelected: false,
      isCorrect: false,
    };
    const { getByText } = render(
      <AdaptivityElementAnswerSelection
        question={{
          questionID: 0,
          questionText: "testQuestionText",
          isMultipleChoice: false,
          difficulty: 0,
          isCompleted: false,
          isRequired: false,
          questionAnswers: [answer],
          hints: [],
        }}
        setHeaderText={() => {}}
        submitSelection={() => {}}
        closeSelection={() => {}}
      />,
    );

    for (let i = 0; i < 2; i++) {
      let oldIsSelected = answer.isSelected;
      getByText("testAnswerText").click();

      expect(answer.isSelected).toBe(!oldIsSelected);
    }
  });

  // ANF-ID: [EWE0008]
  test("click on answer button deselects all other answers if question is single choice", () => {
    const answer1 = {
      answerIndex: 0,
      answerText: "testAnswerText1",
      isSelected: false,
    };
    const answer2 = {
      answerIndex: 1,
      answerText: "testAnswerText2",
      isSelected: false,
    };
    const { getByText } = render(
      <AdaptivityElementAnswerSelection
        question={{
          questionID: 0,
          questionText: "testQuestionText",
          isMultipleChoice: false,
          difficulty: 0,
          isCompleted: false,
          isRequired: false,
          questionAnswers: [answer1, answer2],
          hints: [],
        }}
        setHeaderText={() => {}}
        submitSelection={() => {}}
        closeSelection={() => {}}
      />,
    );

    getByText("testAnswerText1").click();

    expect(answer1.isSelected).toBe(true);
    expect(answer2.isSelected).toBe(false);
  });

  // ANF-ID: [ELG0034]
  test("answer button is green if answer is correct and question is already completed", () => {
    const { container } = render(
      <AdaptivityElementAnswerSelection
        question={{
          questionID: 0,
          questionText: "testQuestionText",
          isMultipleChoice: false,
          difficulty: 0,
          isCompleted: true,
          isRequired: false,
          questionAnswers: [
            {
              answerIndex: 0,
              answerText: "testAnswerText",
              isSelected: true,
              isCorrect: true,
            },
          ],
          hints: [],
        }}
        setHeaderText={() => {}}
        submitSelection={() => {}}
        closeSelection={() => {}}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
