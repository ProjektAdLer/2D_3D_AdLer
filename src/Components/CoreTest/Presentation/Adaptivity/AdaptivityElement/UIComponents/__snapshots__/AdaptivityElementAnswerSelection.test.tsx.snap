// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdaptivityElementAnswerSelection answer button is green if answer is correct and question is already completed 1`] = `
<div>
  <div
    class="flex flex-col w-full p-2 m-auto lg:justify-between lg:flex-row"
  >
    <div
      class="flex flex-wrap justify-start gap-4"
    >
      <div
        class=""
      >
        <button
          class="px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer hover:border-adlerdarkblue hover:bg-adleryellow lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-adlergreen"
        >
          <p
            class="text-sm"
          >
            testAnswerText
          </p>
        </button>
      </div>
    </div>
    <div
      class="flex items-end justify-end w-auto gap-2 pt-8 font-bold lg:w-1/6"
    >
      <div
        class="fixed"
      >
        <div
          class=""
        >
          <button
            class="px-2 py-1 justify-center box-border text-adlerdeactivatedtext bg-adlerbuttonlocked flex items-center text-sm rounded-lg lg:text-xl font-regular border-t-[1px] border-l-[1px] border-b-4 border-r-4 border-white overflow-hidden"
            disabled=""
          >
            <p
              class="text-sm"
            >
              questionAlreadyAnswered
            </p>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AdaptivityElementAnswerSelection should render 1`] = `
<div>
  <div
    class="flex flex-col w-full p-2 m-auto lg:justify-between lg:flex-row"
  >
    <div
      class="flex flex-wrap justify-start gap-4"
    >
      <div
        class=""
      >
        <button
          class="px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer hover:border-adlerdarkblue hover:bg-adleryellow lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue"
        >
          <p
            class="text-sm"
          >
            testAnswerText
          </p>
        </button>
      </div>
    </div>
    <div
      class="flex items-end justify-end w-auto gap-2 pt-8 font-bold lg:w-1/6"
    >
      <div
        class="fixed"
      >
        <div
          class=""
        >
          <button
            class="px-2 py-1 justify-center box-border text-adlerdeactivatedtext bg-adlerbuttonlocked flex items-center text-sm rounded-lg lg:text-xl font-regular border-t-[1px] border-l-[1px] border-b-4 border-r-4 border-white overflow-hidden"
            disabled=""
          >
            <p
              class="text-sm"
            >
              submitAnswer
            </p>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
